### ISSUE TITLE
#![allow(unexpected_cfgs)] does not work for modules

### ISSUE THREAD

**Issue by taiki-e** [rust-lang/rust#124735]
I tried this code:

```rust
// lib.rs
mod a;
```
```rs
// a.rs
#![allow(unexpected_cfgs)] // for cfg(rustfmt)
#![cfg_attr(rustfmt, rustfmt::skip)]
```

I expected to see this happen: no warning

Instead, this happened: `#![allow(unexpected_cfgs)]` is ignored:

```
warning: unexpected `cfg` condition name: `rustfmt`
 --> src/a.rs:2:13
  |
2 | #![cfg_attr(rustfmt, rustfmt::skip)]
  |             ^^^^^^^
  |
  = help: expected names are: `clippy`, `debug_assertions`, `doc`, `docsrs`, `doctest`, `feature`, `miri`, `overflow_checks`, `panic`, `proc_macro`, `relocation_model`, `sanitize`, `sanitizer_cfi_generalize_pointers`, `sanitizer_cfi_normalize_integers`, `target_abi`, `target_arch`, `target_endian`, `target_env`, `target_family`, `target_feature`, `target_has_atomic`, `target_has_atomic_equal_alignment`, `target_has_atomic_load_store`, `target_os`, `target_pointer_width`, `target_thread_local`, `target_vendor`, `test`, `ub_checks`, `unix`, `windows`
  = help: consider using a Cargo feature instead or adding `println!("cargo::rustc-check-cfg=cfg(rustfmt)");` to the top of the `build.rs`
  = note: see <https://doc.rust-lang.org/nightly/cargo/reference/build-scripts.html#rustc-check-cfg> for more information about checking conditional configuration
  = note: `#[warn(unexpected_cfgs)]` on by default
```

Even if adding `#[allow(unexpected_cfgs)]` to `mod a;`, the result is the same.

cc @Urgau

### Meta

`rustc --version --verbose`:
```
rustc 1.80.0-nightly (e82c861d7 2024-05-04)
binary: rustc
commit-hash: e82c861d7e5ecd766cb0dab0bf622445dec999dc
commit-date: 2024-05-04
host: aarch64-apple-darwin
release: 1.80.0-nightly
LLVM version: 18.1.4
```

**Comment by Urgau** [comment]
I would also expect the `#[allow]` to work here.

However, I don't think you need the `cfg_attr` anymore, it can be removed safely.
```diff
- #![cfg_attr(rustfmt, rustfmt::skip)]
+ #![rustfmt::skip]
```

**Comment by taiki-e** [comment]
> it can be removed safely.

No. Custom inner attributes are unstable.

```
error[E0658]: custom inner attributes are unstable
 --> src/lib.rs:1:4
  |
1 | #![rustfmt::skip]
  |    ^^^^^^^^^^^^^
  |
  = note: see issue #54726 <https://github.com/rust-lang/rust/issues/54726> for more information
  = help: add `#![feature(custom_inner_attributes)]` to the crate attributes to enable
  = note: this compiler was built on 2024-05-04; consider upgrading it if it is out of date
```

[playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=20caf0bb074c70a5c6cb9db5ddb3820c)

**Comment by Urgau** [comment]
Oh, good point, but they can (stably) apply it directly to the `mod` like this:
```rust
#[rustfmt::skip]
mod a;
```

**Comment by Urgau** [comment]
Another thing that isn't right here is that the `rustfmt` cfg is not a well known cfg.

I have never been able to find where that cfg might be set by `rustfmt` so I (wrongly) assumed it was anymore, looks this I was wrong.

@rustbot claim

**Comment by taiki-e** [comment]
> they can (stably) be applied directly to the `mod`, like this:
> 
> ```rust
> #[rustfmt::skip]
> mod a;
> ```

Unfortunately, that is not an option available to me due to a rust-analyze bug: https://github.com/rust-lang/rust-analyzer/issues/10826#issuecomment-1647890621

**Comment by lukas-code** [comment]
> I have never been able to find where that cfg might be set by `rustfmt` so I (wrongly) assumed it was anymore, looks this I was wrong.

It looks like rustfmt will just always treat `cfg_attr(anything, rustfmt::skip)` as `rustfmt::skip`, regardless of the actual cfg condition, so this will also work: [(playgound)](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7e091e2d4ce2269ff7e22b29fc26e17b)
```rust
mod m {
    #![cfg_attr(any(), rustfmt::skip)]
    
    fn dont_format_me ()  {   }
}
```

**Comment by fmease** [comment]
> It looks like rustfmt will just always treat cfg_attr(anything, rustfmt::skip) as rustfmt::skip, regardless of the actual cfg condition

Oh, no! >.<

**Comment by Urgau** [comment]
The `rustfmt` cfg is now (as of nighlty-2024-05-07) recognized has a well known cfg.
Fixing one of the issue reported by this issue.

**Comment by jschwe** [comment]
I ran into this issue when trying to disable ASAN instrumentation, without breaking stable users (in an attempt to fix https://github.com/dtolnay/inventory/issues/90). Since ASAN is unstable, and detecting if ASAN is enabled is also unstable, the easiest solution is to require users to set a custom cfg. 

```rust
#[allow(unexpected_cfgs)]
#[cfg_attr(asan, sanitize(address = "off"))]
static MY_ITEM: () = ();
```

```rust
mod my_mod {
    #![allow(unexpected_cfgs)]

    #[cfg_attr(asan, sanitize(address = "off"))]
    static MY_ITEM: () = ();
}
```

The code in question is intended to be generated by a macro in user crates, so requiring all (existing) users to silence the lint, just so that ASAN works again is maybe too much to ask and it would be preferable if the macro could silence the lint, but it appears there is no way to do that.

**Comment by Urgau** [comment]
@jschwe the issue you're describe is a bit different than the problem here. I've opened #155118 to track it.
